File: samples/java-gradle-service/src/main/java/com/example/pricing/PriceCalculator.java
package com.example.pricing;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;

/** Depends on the DiscountRule abstraction, injected through the constructor (DIP). */
public final class PriceCalculator {

    private final List<DiscountRule> rules;
    private final BigDecimal maxDiscountRate;

    public PriceCalculator(List<DiscountRule> rules, BigDecimal maxDiscountRate) {
        this.rules = List.copyOf(rules);
        this.maxDiscountRate = maxDiscountRate;
    }

    public BigDecimal total(Order order) {
        BigDecimal discount = rules.stream()
                .map(rule -> rule.discount(order))
                .reduce(BigDecimal.ZERO, BigDecimal::add);
        BigDecimal cap = order.subtotal().multiply(maxDiscountRate);
        return order.subtotal().subtract(discount.min(cap)).setScale(2, RoundingMode.HALF_UP);
    }
}

File: samples/dotnet-service/src/Pricing/PriceCalculator.cs
namespace Pricing;

/// <summary>Depends on the <see cref="IDiscountRule"/> abstraction, injected via the constructor (DIP).</summary>
/// <param name="rules">Discount rules to apply.</param>
/// <param name="maxDiscountRate">Cap on the total discount, e.g. 0.30 for 30%.</param>
public sealed class PriceCalculator(IEnumerable<IDiscountRule> rules, decimal maxDiscountRate)
{
    private readonly IReadOnlyList<IDiscountRule> _rules = rules.ToList();

    /// <summary>Returns the price after discounts, rounded to cents.</summary>
    public decimal Total(Order order)
    {
        ArgumentNullException.ThrowIfNull(order);
        var discount = _rules.Sum(rule => rule.Discount(order));
        var cap = order.Subtotal * maxDiscountRate;
        return Math.Round(order.Subtotal - Math.Min(discount, cap), 2, MidpointRounding.AwayFromZero);
    }
}
